Clarifications on connection migration (#172)
quicwg/multipath: draft-ietf-quic-multipath.md

Once the implementation has decided which paths to keep alive, it can do so by sending Ping frames
on each of these paths before the idle timeout expires.

## Migration

> 

Reviewer: Migration in RFC 9000 means "move from sending and receiving on path 1 to sending and receiving on path 2". I think that multipath implementation just replaces the RFC 9000 concept to "migration", replacing it by a multiphase process in which path 2 is explicitly tried and validated, and then path 1 might be kept going, or might be explicitly abandoned if not useful any more.

We are discussing here two exceptions to that process: NAT Rebinding, and CID renewal. Mabe we should make that into two sections, instead of a single migration section.

Author: I kept it in one section but am not talking about migration anymore but CID changes and NAT rebinding instead. Better?

Even when the multipath extension is used, migration can happen.
E.g. the client's source port and/or IP address may change due to changes
in network topology or address mappings, such as caused by NAT rebinding.
In this case the server performs path validation (see {{Section 9 of QUIC-TRANSPORT}}).

Moreover, {{Section 5.1.2 of QUIC-TRANSPORT}} indicates that an endpoint
can change the connection ID it uses for a peer to another available one
at any time during the connection. As such a sole change of the Connection

> ID without any change in the address does indiacte a path. 

Author: ```suggestion
ID without any change in the address does not indicate a new change. 
```

If an endpoint uses a new Connection ID after an idle period
and a NAT binding leads to a 4-tuple changes on the same packet,
the receiving endpoint may not be able to associate the packet to
an existing path and will therefore consider this as a new path.
This leads to an inconsistent view of open paths at both peers,
however, as the "old" path will not work anymore, it will be silently

> closed after the idle timeout expires. 

Author: I updated the text but I'm not really happy here. The server has to assume a new path if it can't associate a packet with new CID and new 4-tuple to an existing path but it might still try to use the old path to send data. Those packets will then be lost and it takes a timeout to detect that this is a the case. Isn't there a way for the client to say: ups I received a path validation even though I din't know my address change, so please don't send data using this old CID anymore because it's actually the same path...?

> ## Migration
> 
> Even when the multipath extension is used, migration can happen.
> E.g. the client's source port and/or IP address may change due to changes
> in network topology or address mappings, such as caused by NAT rebinding.
> In this case the server performs path validation (see {{Section 9 of QUIC-TRANSPORT}}).
> 
> Moreover, {{Section 5.1.2 of QUIC-TRANSPORT}} indicates that an endpoint
> can change the connection ID it uses for a peer to another available one
> at any time during the connection. As such a sole change of the Connection
> ID without any change in the address does indiacte a path. 
> 
> If an endpoint uses a new Connection ID after an idle period
> and a NAT binding leads to a 4-tuple changes on the same packet,
> the receiving endpoint may not be able to associate the packet to
> an existing path and will therefore consider this as a new path.
> This leads to an inconsistent view of open paths at both peers,
> however, as the "old" path will not work anymore, it will be silently
> closed after the idle timeout expires. 
> 
> {{Section 9.3 of QUIC-TRANSPORT}} allows an endpoint to skip validation of
> a peer address if that address has been seen recently. However, when the
> multipath extension is used and an endpoint has multiple addresses that
> could lead to switching between different paths, it should rather maintain
> multiple open paths instead.
> 
> 

Reviewer 2: I wonder if the following would make it more clear by avoiding the notion of "migration".

Note that I think we could have a better behaviour in the case "both 4-tuple and CID change", but I would prefer delegating this to a later issue.
```suggestion
## NAT Rebinding {nat-rebinding}

While endpoints assign a Connection ID to a specific sending 4-tuple,
networks events such as NAT rebinding may make the packet's receiver
perceives a different 4-tuple. Servers observing 4-tuple changes must
initiate path validation. If the path validation process succeeds, the
endpoint follows considerations of {{Section 9.4 of QUIC-TRANSPORT}}.

## CID Renewal

At any time, endpoints may change the Connection ID they use to send
packets on a path. Such a path can keep the same congestion control
state and RTT measurements.

If an endpoint uses a new Connection ID after an idle period
and a NAT rebinding leads to a 4-tuple changes on the same packet,
the receiving endpoint may not be able to associate the packet to
an existing path and will therefore consider this as a new path.
This leads to an inconsistent view of open paths at both peers,
however, as the "old" path will not work anymore, it will be silently
closed after the idle timeout expires.

```

Author: Sorry I did another change and lost you proposal but will edit the PR directly now accordingly. Thx!

Once the implementation has decided which paths to keep alive, it can do so by sending Ping frames
on each of these paths before the idle timeout expires.

## Connection ID Changes and NAT Rebindings

{{Section 5.1.2 of QUIC-TRANSPORT}} indicates that an endpoint

> can change the connection ID it uses for to another available one

Author: ```suggestion
can change the Connection ID it uses for to another available one
```

at any time during the connection. As such a sole change of the Connection
ID without any change in the address does not indicate a path change and

> the endpoint will keep the same congestion control and RTT measurement state.

Reviewer 2: Given that we give implementation considerations, we should not restrict the behaviour too much IMHO.

```suggestion
the endpoint can keep the same congestion control and RTT measurement state.
```

While endpoints assign a Connection ID to a specific sending 4-tuple,
networks events such as NAT rebinding may make the packet's receiver
observe a different 4-tuple. Servers observing a 4-tuple change will
performs path validation (see {{Section 9 of QUIC-TRANSPORT}}).
If the path validation process succeeds, the endpoints need to reset

> their congestion controller and round-trip time estimator for he new

Author: ```suggestion
their congestion controller and round-trip time estimator for the new
```

> If the path validation process succeeds, the endpoints need to reset
> their congestion controller and round-trip time estimator for the new
> path as specficied in  {{Section 9.4 of QUIC-TRANSPORT}}.

Reviewer 2: Why not simply refers to that section directly, that also states that servers may keep their current state if they suspect the event to be a NAT rebinding?
```suggestion
If the path validation process succeeds, the endpoint
follows considerations of {{Section 9.4 of QUIC-TRANSPORT}}.
```

Author: I would prefer to keep this because I find it annoying to provide a vague reference without giving an indication what this reference is for. Why do you think the current text is not good?

Reviewer 2: Because it somewhat contradicts what Section 9.4 states.

> On confirming a peer's ownership of its new address, an endpoint MUST immediately reset the congestion controller and round-trip time estimator for the new path to initial values (see Appendices [A.3](https://www.rfc-editor.org/rfc/rfc9002#appendix-A.3) and [B.3](https://www.rfc-editor.org/rfc/rfc9002#appendix-B.3) of [[QUIC-RECOVERY](https://www.rfc-editor.org/rfc/rfc9000.html#QUIC-RECOVERY)]) unless the only change in the peer's address is its port number. Because port-only changes are commonly the result of NAT rebinding or other middlebox activity, the endpoint MAY instead retain its congestion control state and round-trip estimate in those cases instead of reverting to initial values.

We could just adapt the text to be like "the endpoint updates the path's congestion controller and round-trip time estimator according to {{Section 9.4 of QUIC-TRANSPORT}}".

Author: ```suggestion
If path validation process succeeds, the endpoints set
the path's congestion controller and round-trip time
estimator according to {{Section 9.4 of QUIC-TRANSPORT}}.
```

{{Section 9.3 of QUIC-TRANSPORT}} allows an endpoint to skip validation of
a peer address if that address has been seen recently. However, when the
multipath extension is used and an endpoint has multiple addresses that
could lead to switching between different paths, it should rather maintain
multiple open paths instead.

If an endpoint uses a new Connection ID after an idle period
and a NAT rebinding leads to a 4-tuple changes on the same packet,
the receiving endpoint may not be able to associate the packet to
an existing path and will therefore consider this as a new path.
This leads to an inconsistent view of open paths at both peers,
however, as the "old" path will not work anymore, it will be silently

> closed after the idle timeout expires. 

Author: ```suggestion
closed after the idle timeout expires.
```

> 

Reviewer: I think overloading the meaning of PATH_CHALLENGE is a bad idea, see comments on #188. But sending a "RETIRE_CONNECTION_ID" too soon is also a bad idea, because the peer needs some time to send acknowledgements. Maybe the peer doing the CID renewal could send PATH_ABANDON(old_path, reason_code = CID_replaced), with a clear statement explaining that the 4-tuple is still valid, but the CID is not?

> 

Author: Let's open an issue to add a reason code, or a replaced by field, or both. Maybe we should also rename path_abandon to cid_abandon?

Question is should we merge this text for now or also open a new issue instead? I think we can merge because it shouldn't hurt to send a path_abandon frame, right?

> 

Reviewer 2: I think this issue is somewhat orthogonal, and discussed in #188.

> 

Reviewer: We do have an issue with the server continuing sending data on the old path, now a dead end. I am concerned that we will not have an agreement before Monday. I left comments in issue #188. If we resolve that issue before Friday 3/10, let's incorporate the agreed solution in this PR. If we do not, let's check in the PR as is, without the extra text.

Once the implementation has decided which paths to keep alive, it can do so by sending Ping frames
on each of these paths before the idle timeout expires.

## Migration

Even when the multipath extension is used, migration can happen.
E.g. the client's source port and/or IP address may change due to changes
in network topology or address mappings, such as caused by NAT rebinding.
In this case the server performs path validation (see {{Section 9 of QUIC-TRANSPORT}}

> and both peers are expected to use new connection IDs during this process.

Author: My thinking was that if you receive a path challenge from the server, you should treat this as a "request" to open a new path and thus use a new CID even if the 4-tuple seem to correspond to an existing path.

However, thinking further I guess you also have to close the existing path at the same time as you should not have two paths with the same 4-tuple. Thus the client MUST send a path abandon frame for the old path.

Even is the path validation is causes by a migration event, when
the multipath extension is used, the
endpoint treats the receipt of a PATH_CHALLENGE frame as a new path

> and also initiates path validation to open the path.

Author: I think this is something we need to decide.

The problem is that the server sees a new 4-tuple but the client not, so we need to make clear if this creates a new path or not in such a way that both ends come to the same conclusion.

You could also say it's not a new path if on server side the 4-tuple changes but not the CID and on client side a path challenge (with new CID) is received on an existing 4-tuple. However, you still need to reset congestion control in most cases, so I thought it was easier to just create a new path.

> and also initiates path validation to open the path.

Author: And you still need to address the case where both the CID and the 4-tuple changes. As you can't decide to which existing path you should assign this packet, I guess you have to assume it's a new path. Therefore it better if the client also opens a new path when it received the path challenge and explicitly closes the old path.

This is important to ensure

> that both peers have the same view on the number of active paths. However,

Author: If both the 4-tuple and the CID changes, the client should send a path challenge in that packet to explicitly open a new path. Otherwise the server should probably ignore that packet as it can't associate it to an exiting path. Maybe something we should also spell out more clearly.

> that both peers have the same view on the number of active paths. However,

Author: Actually ignoring might not be right to address this change where you restart after an idle period with a new CID and then also the NAT rebinding happens... however, it would be the same procedure as in the case when the CID not changes after a NAT rebinding: the server sends a path challenge frame and that would be the signal for the client to assign a new path and use a new CID + close the old path on the presumedly same 4-tuple from the client's point of view.

> that both peers have the same view on the number of active paths. However,

Reviewer: The node that renewed the CID will deduce the path change from invoming traffic on same (incoming) CID and new 4-tuple. It can at any point abandon the old path or retire the old CID.

> that both peers have the same view on the number of active paths. However,

Author: The problem is that after removing the type we used the sender’s cid in the abandon frame and then this doesn’t work as the client only sees one path. However I think we need to fix that an then recommend to send an abandon frame for the old received cid.

this also implicitly means that every migration event will cause a congestion
control reset, even if e.g. only the source port changed. Further, if the
old path after a migration event is not usable anymore, with the multipath
extension the endpoint might send a PATH_ABANDON frame to close the path.

{{Section 9.3 of QUIC-TRANSPORT}} allows an endpoint to skip validation of
a peer address if that address has been seen recently. However, when the
multipath extension is used and an endpoint has multiple addresses that
could lead to switching between different paths, it should rather maintain
multiple open paths instead. Otherwise switching paths without path validation
could cause an inconsistent view of open paths at both peer.

Moreover, {{Section 5.1.2 of QUIC-TRANSPORT}} indicates that an endpoint
can change the connection ID it uses for a peer to another available one
at any time during the connection. As such a sole change of the Connection
ID without any change in the address does open a new path.

> 

Author: First, I don't think it is true that this is not aligned with RFC9000 because there is always just one path in RFC9000. However, the problem is that in RFC9000 both ends don't need a common understanding of the path but now they do, so we need to address this. In our definition in this draft every 4-tuple is a different path and sticking to that makes it less complicated. The problem with NAT rebinding is that only one end see a new 4-tuple. As such we need an explicit signal to tell the other end to create a new path. I propose to simple use path validation for this (to keep change minimal).

If you are concerned about resetting the congestion window, we can say that the new path can be initialisierend with the cc values from the old path if only the source port changes.
